src/discord/discord.py

```diff
@@ -0,0 +1,277 @@
+    result = await _discord_request(url, tok)
+
+    json_s = result.json()
+    users = [GuildMember(User(user['user']['id'], user['user']['username'], user['user']['discriminator'], user['user']['global_name'], user['user']['avatar']), user['roles']) for user in json_s]
+    last_uid = users[-1].user.id
+
+    while True:
+        url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000&after={last_uid}'
+        result = await _discord_request(url, tok)
+
+        json_s = result.json()
+        res = [GuildMember(User(user['user']['id'], user['user']['username'], user['user']['discriminator'], user['user']['global_name'], user['user']['avatar']), user['roles']) for user in json_s]
+        users += res
+
+        if res == []:
+            return users
+
+        last_uid = res[-1].user.id
+
+async def get_categories(
+    id: str = guild_id
+) -> list[str]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/channels'
+    result = await _discord_request(url, tok)
+
+    result_json = result.json()
+    return [category['name'] for category in result_json if category['type'] == DISCORD_CATEGORY_ID]
+
+async def get_channels_by_category_name(
+    category_name: str,
+    id: str = guild_id
+) -> list[Channel]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/channels'
+    result = await _discord_request(url, tok)
+
+    result_json = result.json()
+    # TODO: edge case if there exist duplicate category names, see get_channels_by_category_id()
+    category_id = [category['id'] for category in result_json if category['type'] == DISCORD_CATEGORY_ID and category['name'] == category_name][0]
+    channels = [Channel(channel['id'], channel['type'], channel['guild_id'], channel['name'], channel['permission_overwrites']) 
+                for channel in result_json if channel['type'] != DISCORD_CATEGORY_ID and channel['parent_id'] == category_id]
+    return channels
+
+async def get_channels_by_category_id(
+    cid: str,
+    id: str = guild_id
+) -> list[Channel]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/channels'
+    result = await _discord_request(url, tok)
+
+    result_json = result.json()
+    channels = [Channel(channel['id'], channel['type'], channel['guild_id'], channel['name'], channel['permission_overwrites']) 
+                for channel in result_json if channel['type'] != DISCORD_CATEGORY_ID and channel['parent_id'] == cid]
+    return channels
+
+async def search_user(
+    user: str,
+    id: str = guild_id
+) -> User:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/members/search?query={user}'
+    result = await _discord_request(url, tok)
+    json = result.json()[0]['user']
+    user = User(json['id'], json['username'], json['discriminator'], json['global_name'], json['avatar'])
+    return user
```

Comment:
We should avoid shadowing the user parameter, as it could be confusing.
```suggestion
    return User(json['id'], json['username'], json['discriminator'], json['global_name'], json['avatar'])
```